x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -338,84 +358,172 @@ public void removeFromCache(final CacheKey cacheKey) {
+                logger.trace("shard eviction [{}] is being processed", shardEviction);
+                return runningShardsEvictions.get(shardEviction);
+
+            } else {
+                completableFuture = new CompletableFuture<>();
+                runningShardsEvictions.put(shardEviction, completableFuture);
+                completableFuture.whenComplete((result, e) -> {
+                    synchronized (shardsEvictionsMutex) {
+                        final CompletableFuture<Integer> removedFuture = runningShardsEvictions.remove(shardEviction);
+                        final boolean removed = pendingShardsEvictions.remove(shardEviction);
+                        assert removedFuture == completableFuture;
+                        assert removed;
+                        assert shardsEvictionsInvariant();
+                    }
+                });
+            }
+        }
+        try {
+            final List<CacheFile> cacheFilesToEvict = new ArrayList<>();
+            cache.forEach((cacheKey, cacheFile) -> {
+                if (shardEviction.matches(cacheKey)) {
+                    cacheFilesToEvict.add(cacheFile);
                 }
-                success = true;
-            } finally {
-                assert success : "shard eviction should be successful: " + shardEviction;
-                if (success == false) {
-                    final boolean added = evictedShards.add(shardEviction);
-                    assert added : shardEviction;
+            });
+            for (CacheFile cacheFile : cacheFilesToEvict) {
+                try {
+                    cache.invalidate(cacheFile.getCacheKey(), cacheFile);
+                } catch (RuntimeException e) {
+                    logger.warn(() -> new ParameterizedMessage("failed to evict cache file {}", cacheFile.getCacheKey()), e);
+                    assert false : e;
                 }
             }
+            logger.debug("shard eviction [{}] processed with [{}] cache files invalidated", shardEviction, cacheFilesToEvict.size());
+            completableFuture.complete(cacheFilesToEvict.size());
+        } catch (Exception e) {
+            completableFuture.completeExceptionally(e);
         }
+        return completableFuture;
+    }
+
+    /**
+     * Processes and waits for all pending shard evictions to complete.
+     */
+    private void processAllPendingShardsEvictions() {
+        final List<CompletableFuture<Integer>> pendingFutures;
+        synchronized (shardsEvictionsMutex) {
+            allowShardsEvictions = false;
+            pendingFutures = new ArrayList<>(runningShardsEvictions.values());
+            assert shardsEvictionsInvariant();
+        }
+        if (pendingFutures.isEmpty() == false) {
+            try {
+                final CountDownLatch latch = new CountDownLatch(pendingFutures.size());
+                pendingFutures.forEach(completableFuture -> completableFuture.whenComplete((integer, throwable) -> latch.countDown()));
```

Comment:
I think it is ok to only wait for running shard evictions to complete and not process other shard evictions that have not been processed yet: if the shard files are still on disk they will be reused or removed after node starts again and if the shard files have been deleted from disk they won't be reloaded at node start up by the persistent cache logic.

Reply:
Agreed. We want to terminate as fast as we can.

I wonder if a simple read-write lock scheme could suffice? Processing the eviction takes the read-lock (and checks that state==started) and stop takes the write lock.